contracts/BC_fusion/StakeHub.sol

```diff
@@ -0,0 +1,786 @@
+    }
+
+    function delegate(address operatorAddress) external payable whenNotPaused validatorExist(operatorAddress) {
+        uint256 bnbAmount = msg.value;
+        require(bnbAmount >= minDelegationBNBChange, "INVALID_DELEGATION_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory valInfo = _validators[operatorAddress];
+
+        if (valInfo.jailed) {
+            // only self delegation
+            require(delegator == operatorAddress, "ONLY_SELF_DELEGATION");
+        }
+
+        uint256 shares = IStakeCredit(valInfo.creditContract).delegate{ value: bnbAmount }(delegator);
+        emit Delegated(operatorAddress, delegator, shares, bnbAmount);
+
+        _sync(valInfo.creditContract, delegator);
+        IGovToken(GOV_TOKEN_ADDR).delegateVote(delegator, operatorAddress);
+    }
+
+    function undelegate(address operatorAddress, uint256 shares) public whenNotPaused validatorExist(operatorAddress) {
+        require(shares > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory valInfo = _validators[operatorAddress];
+        uint256 bnbAmount = IStakeCredit(valInfo.creditContract).undelegate(delegator, shares);
+        emit Undelegated(operatorAddress, delegator, shares, bnbAmount);
+
+        if (delegator == operatorAddress) {
+            _checkValidatorSelfDelegation(operatorAddress);
+        }
+
+        _sync(valInfo.creditContract, delegator);
+    }
+
+    function redelegate(
+        address srcValidator,
+        address dstValidator,
+        uint256 shares
+    ) public whenNotPaused validatorExist(srcValidator) validatorExist(dstValidator) validatorNotJailed(dstValidator) {
+        require(shares > 0, "INVALID_SHARES_AMOUNT");
+
+        address delegator = msg.sender;
+        Validator memory srcValInfo = _validators[srcValidator];
+        Validator memory dstValInfo = _validators[dstValidator];
+        require(
+            IStakeCredit(srcValInfo.creditContract).getPooledBNBByShares(shares) >= minDelegationBNBChange,
+            "INVALID_REDELEGATION_AMOUNT"
+        );
+
+        if (dstValInfo.jailed) {
+            // only self delegation
+            require(delegator == dstValidator, "ONLY_SELF_DELEGATION");
+        }
+
+        uint256 bnbAmount = IStakeCredit(srcValInfo.creditContract).unbond(delegator, shares);
+        uint256 newShares = IStakeCredit(dstValInfo.creditContract).delegate{ value: bnbAmount }(delegator);
+        emit Redelegated(srcValidator, dstValidator, delegator, shares, newShares, bnbAmount);
+
+        if (delegator == srcValidator) {
+            _checkValidatorSelfDelegation(srcValidator);
+        }
+
+        address[] memory _pools = new address[](2);
+        _pools[0] = srcValInfo.creditContract;
+        _pools[1] = dstValInfo.creditContract;
+        IGovToken(GOV_TOKEN_ADDR).sync(_pools, delegator);
+        IGovToken(GOV_TOKEN_ADDR).delegateVote(delegator, dstValidator);
```

Comment:
so is here, it happens only when user want it happen

Reply:
Done